ledger-tool/src/main.rs

```diff
@@ -4167,59 +4171,66 @@ fn main() {
-                                    last
-                                );
-                                if all {
-                                    println!("Non-empty slots: {slots:?}");
-                                }
-                            } else {
-                                println!("Ledger has data for slot {first:?}");
-                            }
-                        }
-                        if let Ok(rooted) = blockstore.rooted_slot_iterator(0) {
-                            let mut first_rooted = 0;
-                            let mut last_rooted = 0;
-                            let mut total_rooted = 0;
-                            for (i, slot) in rooted.into_iter().enumerate() {
-                                if i == 0 {
-                                    first_rooted = slot;
-                                }
-                                last_rooted = slot;
-                                total_rooted += 1;
+                            // Collect info about slot bounds
+                            let mut bounds = SlotBounds {
+                                slots: SlotInfo {
+                                    total: slots.len(),
+                                    first: Some(*slots.first().unwrap()),
+                                    last: Some(*slots.last().unwrap()),
+                                    ..SlotInfo::default()
+                                },
+                                ..SlotBounds::default()
+                            };
+                            if all {
+                                bounds.all_slots = Some(&slots);
                             }
-                            let mut count_past_root = 0;
-                            for slot in slots.iter().rev() {
-                                if *slot > last_rooted {
-                                    count_past_root += 1;
-                                } else {
-                                    break;
+
+                            // Consider also rooted slots, if present
+                            if let Ok(rooted) = blockstore.rooted_slot_iterator(0) {
+                                let mut first_rooted = 0;
+                                let mut last_rooted = 0;
+                                let mut total_rooted = 0;
+                                for (i, slot) in rooted.into_iter().enumerate() {
+                                    if i == 0 {
+                                        first_rooted = slot;
+                                    }
+                                    last_rooted = slot;
+                                    total_rooted += 1;
                                 }
+
+                                let count_past_root = slots
+                                    .iter()
+                                    .rev()
+                                    .take_while(|slot| *slot > &last_rooted)
+                                    .count();
```

Comment:
```suggestion
                                let mut first_rooted = None;
                                let mut last_rooted = None;
                                let mut total_rooted = 0;
                                for (i, slot) in rooted.into_iter().enumerate() {
                                    if i == 0 {
                                        first_rooted = Some(slot);
                                    }
                                    last_rooted = Some(slot);
                                    total_rooted += 1;
                                }

                                let last_root_for_comparison = last_rooted.unwrap_or_default();
                                let count_past_root = slots
                                    .iter()
                                    .rev()
                                    .take_while(|slot| *slot > &last_root_for_comparison)
                                    .count();
```
Prior to this change, we might get output like:
```
{
  "slots": {
    "total": 101,
    "first": 171368538,
    "last": 171368638
  },
  "roots": {
    "total": 0,
    "first": 0,
    "last": 0,
    "numAfterLastRoot": 101
  }
}
```
Notice that the first/last root are both 0; this is because we were setting them to Some(0) even if there weren't any roots. So, this change will make the json output display `null` if there are no roots:
```
{
  "slots": {
    "total": 101,
    "first": 171368538,
    "last": 171368638
  },
  "roots": {
    "total": 0,
    "first": null,
    "last": null,
    "numAfterLastRoot": 101
  }
}
```

Reply:
That's perfectly clear :) I'll update right now